docs/authentication.md

```diff
@@ -0,0 +1,318 @@
+
+```jsx
+// pages/api/login.js
+
+import auth0 from '../../lib/auth0';
+
+export default async function login(req, res) {
+  try {
+    await auth0.handleLogin(req, res);
+  } catch (error) {
+    console.error(error);
+    res.status(error.status || 500).end(error.message);
+  }
+}
+```
+
+Auth0 populates the session so you can utilize static generation or `getServerSideProps` for server-side rendering. Here's an example of the latter.
+
+```jsx
+// pages/profile.js
+
+import auth0 from '../../lib/auth0';
+
+export async function getServerSideProps({ req, res }) {
+  const session = await auth0.getSession(req);
+
+  if (!session || !session.user) {
+    res.writeHead(307, { Location: '/api/login' });
+    res.end();
+    return { props: {} };
+  }
+
+  return { props: { user: session.user } };
+}
+```
+
+<details open>
+  <summary><b>Examples</b></summary>
+  <ul>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/auth0">auth0</a></li>
+  </ul>
+</details>
+
+## Frequently Asked Questions
+
+### How can I fetch user data once for every page in my application?
+
+To fetch data once for your entire Next.js application, create a [custom `App`](https://nextjs.org/docs/advanced-features/custom-app) and use `getInitialProps`.
+You should only use this method if you have blocking data requirements for _every single page_ in your application.
+This will disable [static optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization).
+We are [working on adding support](https://github.com/vercel/next.js/discussions/10949#discussioncomment-44898) for `getStaticProps` inside `_app.js`.
+
+Here's an example using a provider like Auth0.
+
+```jsx
+// _app.js
+
+App.getInitialProps = async ({ req, res }) => {
+  // Only run on the server
+  if (typeof window === 'undefined') {
+    const session = await auth0.getSession(req);
+    if (!session || !session.user) {
+      res.writeHead(307, { Location: '/api/login' });
+      res.end();
+
+      return;
+    }
+
+    return { user: session.user };
+  }
+};
+```
+
+If you require some pages to render as static HTML, you should use `getServerSideProps`
+in every page requiring server-rendering. You can extract this logic to a shared function to improve code reuse.
+
+```jsx
+// lib/auth.js
+export const requirePageAuth = (inner) => {
+  return async (context) => {
+    const session = await getSession(context.req);
+
+    if (!session) {
+      context.res.writeHead(307, { Location: '/login' });
+      context.res.end();
+      return { props: {} };
+    }
+
+    return inner ? inner(context, auth) : { props: { session } };
+  };
+};
+```
+
+```jsx
+// pages/profile.js
+
+export default Profile = ({ session }) => {
+  return <p>{session.user.name}</p>;
+};
+
+export const getServerSideProps = requirePageAuth;
+```
```

Comment:
This should be under the SSR section. This may be very common and probably not needed under FAQ. (We could also just mention it and link to the example)

Reply:
Good idea, I agree we should move this up. Which example in particular are you referencing?